### ISSUE TITLE
feat(ios): support for databases that are not visible to iTunes and are not backed up by iCloud

### ISSUE THREAD

**Issue by robingenz** [capacitor-community/sqlite#207]
**Is your feature request related to a problem? Please describe.**

The iOS databases are currently stored in the `documentDirectory`, which is visible to iTunes and backed up to iCloud.

**Describe the solution you'd like**

I would like to be able to specify that the database is stored in a location that is not visible to iTunes and not backed up to iCloud.
As far as I know, the `libraryDirectory` is not visible to iTunes. For this reason I suggest to save the databases under the following path: `Library/LocalDatabase`.
To prevent the database from being backed up by iCloud, it is necessary to set `NSURLIsExcludedFromBackupKey` for the folder, i.e. `LocalDatabase`, to `true` (example: https://stackoverflow.com/a/32807331/6731412).

This could be implemented via the plugin configuration to avoid breaking changes.
An example implementation can be found here: https://github.com/robingenz/capacitor-badge#configuration and https://github.com/robingenz/capacitor-badge/blob/b2a4256c77a4cac251104478d007757ef8de21e4/ios/Plugin/BadgePlugin.swift#L114

Config Interface Example `capacitor.config.json`:

```json
{
  "plugins": {
    "Sqlite": {
      "iosDatabaseLocation": "Library/LocalDatabase"
    }
  }
}
```

I'm ready to help with the implementation.

**Comment by jepiqueau** [comment]
@robingenz thanks for making the PR. I think the capacitor databases should be stored in a different folder than others cordova plugins are storing their databases. I will suggest Library/CapacitorDatabase i will have a look at your suggestion for the capacitor.config.json. i wish you an Happy New Year

**Comment by robingenz** [comment]
@jepiqueau Thanks for your quick reply! Please let me know as soon as I can help. I wish you a happy new year too! Keep up the great work!

**Comment by jepiqueau** [comment]
@chriswep Hi Chris i wish you first an Happy New Year.  Can you comment this request and agree on it and tell me the impact on the typeOrm driver

**Comment by jepiqueau** [comment]
@robingenz it will come soon. I have still to update the doc and to look at some others issues before to publish a  new release

**Comment by jepiqueau** [comment]
@robingenz this is fixed in 3.3.3-2 this has been tested in `[react-sqlite-app-starter](https://github.com/jepiqueau/react-sqlite-app-starter)`

**Comment by robingenz** [comment]
@jepiqueau Thank you very much! I'll test it and let you know.

**Comment by robingenz** [comment]
@jepiqueau I am currently encountering a strange problem with version `3.3.3-3`.
I get the following error message when running the db migration:
```
2022-01-03 19:02:02.468814+0100 ENGINE4[987:77715] open on /var/mobile/Containers/Data/Application/768E7DC9-3B74-4382-99EB-3E80BFB58100/Library/CapacitorDatabase/mfapp.globalSQLite.db: No such file or directory
Error: Error Domain=NSCocoaErrorDomain Code=4 "Die Datei „mfapp.global“ existiert nicht." UserInfo={NSSourceFilePathErrorKey=/var/mobile/Containers/Data/Application/768E7DC9-3B74-4382-99EB-3E80BFB58100/Library/LocalDatabase/mfapp.global, NSUserStringVariant=(
    Copy
), NSDestinationFilePath=/var/mobile/Containers/Data/Application/768E7DC9-3B74-4382-99EB-3E80BFB58100/Library/CapacitorDatabase/mfapp.globalSQLite.db, NSFilePath=/var/mobile/Containers/Data/Application/768E7DC9-3B74-4382-99EB-3E80BFB58100/Library/LocalDatabase/mfapp.global, NSUnderlyingError=0x283660f00 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}
Error: copyFile Failed
```

My code: 
```js
private async migrateCordovaSqliteDb(): Promise<void> {
    const folderPath = 'Library/LocalDatabase';
    const result = await this.sqlite.getMigratableDbList(folderPath);
    const dbNames = (result.values || []).filter(
        (value: string) => !!value && value.includes('mfapp.'),
    );
    if (dbNames.length < 1) {
        return;
    }
    await this.sqlite.addSQLiteSuffix(folderPath, dbNames);
    await this.sqlite.deleteOldDatabases(folderPath, dbNames);
}
```

```json
"CapacitorSQLite": {
    "iosDatabaseLocation": "Library/CapacitorDatabase"
}
```

I am currently still debugging and trying to find the reason.

The following file exists: `Library/LocalDatabase/mfapp.global`
I checked this before by downloading the app container in XCode.

It seems that the problem only occurs when the file `CapacitorDatabase/mfapp.globalSQLite.db` does NOT exist.
If I skip the migration first and create the CapacitorSQLite database via createConnection and execute the migration only afterwards, then it works.

Do you have any idea what could be the reason for this?

**Comment by jepiqueau** [comment]
@robingenz Yes the point in the database name is not supported `mfapp.global`

**Comment by robingenz** [comment]
Okay, but it worked in the previous version. I have been using the migration with this names for a few weeks now.
So there was a breaking change in a minor version? Was that intentional?

**Comment by jepiqueau** [comment]
@robingenz OK i will have to look back on this. No it was not intentional. I had to make so many changes and this was not part of m'y tests

**Comment by jepiqueau** [comment]
@robingenz in the `Library/LocalDatabase` the database mfapp.global has not .db as extension?

**Comment by jepiqueau** [comment]
@robingenz if the mfapp.global does not have the extension .db, the .global is taken as the extension  and the not copied

**Comment by jepiqueau** [comment]
@robingenz forget all what i said above i tested it and it works fine 
i add two databases in the `Library/LocalDatabase`
`mfapp.global` , `mfapp.test.db`
one without the db extension and one with the db extension

```js
        const testMigrateIssue = async(): Promise<Boolean>  => {
            setLog((log) => log.concat("* Starting testMigrateIssue *\n"));

            try {
                const folderPath = 'Library/LocalDatabase';
                const result = await sqlite.getMigratableDbList(folderPath);
                const dbNames = (result.values || []).filter(
                    (value: string) => !!value && value.includes('mfapp.'),
                );
                if (dbNames.length < 1) {
                    errMess.current = `NO "mfapp. " database`;
                    return false;
                }
                console.log(`dbNames: ${JSON.stringify(dbNames)}`)
                await sqlite.addSQLiteSuffix(folderPath, dbNames);
                const listDB = await sqlite.getDatabaseList();
                console.log(`listDB: ${JSON.stringify(listDB)}`)
//                await sqlite.deleteOldDatabases(folderPath, dbNames);
            
                return true;
            } catch (err: any) {
                errMess.current = `${err.message}`;
                return false;
            }
```
and the output
```
  To Native ->  CapacitorSQLite addSQLiteSuffix 9265317
⚡️  [log] - dbNames: ["mfapp.global","mfapp.test.db"]
⚡️  TO JS undefined
⚡️  To Native ->  CapacitorSQLite getDatabaseList 9265318
⚡️  TO JS {"values":["testListenersSQLite.db","myDBSQLite.db","db_issue9SQLite.db","testfromfileSQLite.db","mfapp.testSQLite.db","mfapp.globalSQLite.db","NoEncryptionSQLite.db","test-updversionSQLite.db","testSetSQLite.db","stewardSQLite.db","testcopySQLite.db","tes
⚡️  [log] - listDB: {"values":["testListenersSQLite.db","myDBSQLite.db","db_issue9SQLite.db","testfromfileSQLite.db","mfapp.testSQLite.db","mfapp.globalSQLite.db","NoEncryptionSQLite.db","test-updversionSQLite.db","testSetSQLite.db","stewardSQLite.db","testcopySQLite.db","testNewSQLite.db","testListenersImportedSQLite.db","dbForCopySQLite.db"]}

```

**Comment by robingenz** [comment]
> @robingenz in the `Library/LocalDatabase` the database mfapp.global has not .db as extension?

Yes

**Comment by jepiqueau** [comment]
@robingenz for me it is working as shown above

**Comment by jepiqueau** [comment]
@robingenz is it working for you ? if yes can you close the issue

**Comment by robingenz** [comment]
Unfortunately, I didn't have time yesterday. I will test it again today and let you know.
Thanks for your help!

**Comment by jepiqueau** [comment]
@robingenz i will be travelling today so nt accessible. I wish you good luck

**Comment by robingenz** [comment]
@jepiqueau I have found the problem: In my case, the CapacitorDatabase folder was not yet created at the time of migration. 
The reason for this is that `class func copyFile(pathName: String, toPathName: String, overwrite: Bool) throws -> Bool` is called which, unlike `class func copyFile(fileName: String, toFileName: String, databaseLocation: String) throws -> Bool`, does not call the `getFilePath`/`createDirCopyDB` method which creates the folder.

This also fits with my description of the problem, that it works after restarting the app with an empty database.
I added the following code locally to test my assumption:

Update `init` method in `CapacitorSQLite`:

```diff
init(config: SqliteConfig) {
    self.config = config
    if let iosLocation = config.iosDatabaseLocation {
        self.databaseLocation = iosLocation
+        UtilsFile.tryCreateIosDatabaseLocation(location: iosLocation)
    } else {
        self.databaseLocation = "Documents"
    }
    super.init()
}
```

Added `tryCreateIosDatabaseLocation` method to `UtilsFile`:

```swift
class func tryCreateIosDatabaseLocation(location: String) {
    do {
        var dirUrl: URL = try UtilsFile
            .getFolderURL(folderPath: location)
        let dirPath: String = dirUrl.path
        if !UtilsFile.isDirExist(dirPath: dirPath) {
            // create the directory
            try FileManager.default
                .createDirectory(at: dirUrl, withIntermediateDirectories: true)
            // exclude the directory from iCloud Backup
            try UtilsFile.setExcludeFromiCloudBackup(&dirUrl,
                                                     isExcluded: true)
        }
    } catch let error {
        var msg: String = "tryCreateIosDatabaseLocation command failed :"
        msg.append(" \(error.localizedDescription)")
        print(msg)
    }
}
```

With this code change it works again.

Unfortunately this is not a fix yet, as then the `createDirCopyDB` method does not work anymore, as the folder always exists (as it is created directly at startup).

**Comment by jepiqueau** [comment]
@robingenz that is strange as for me it works does not matter if the directory exists or not

**Comment by robingenz** [comment]
@jepiqueau As far as I know, in order to use `FileManager.copyItem` you have to make sure that the directories exist.
So as a fix I would suggest that before copying the database you always check if the folder exists and if not you create it.
Is the `moveAllDBSQLite` method really needed? After all there are already methods for migration (`getMigratableList`, `addSQLiteSuffix`).
Then you could create the folder like in my example directly at the start of the app and avoid such problems from the beginning.

**Comment by jepiqueau** [comment]
@robingenz OK i will have a look

**Comment by jepiqueau** [comment]
@robingenz You are right, i did not see it in my test app as i create a database earlier in the `index.tsx` file. So when i tested the addSQLiteSuffix the `Library/CapacitorDatabase` was created. So i will fix it. Thanks for your patience and temerity

**Comment by robingenz** [comment]
Great!
Thank you very much for your efforts.

**Comment by jepiqueau** [comment]
@robingenz fix in 3.3.3-5

**Comment by robingenz** [comment]
@jepiqueau Thank you!
